### PR TITLE
Bugfix/non ascii in xml

### PR DESCRIPTION
**Pull request by shino**
This PR addresses two issues #787 and #628.

The plan to fix them and possible other XML-with-multibyte(non-ASCII)-character
issues are as follows:
- `riak_cs_xml:format_value(Val)` treats conversion between lists/binaries and
  Unicode strings
  (xmerl accepts only Unicode strings but not binary or latin-1-ish strings).
- In the function clause `when is_list(Val)`, use list_to_binary and call
  `format_value` again with converted binary.
- In the function clause `when is_binary(Val)` use `unicode:characters_to_list`
  to produce Unicode strings.
- Remove all other `unicode:characters_to_list` around XML output creation.

Some misc notes:
- When we take into account Unicode, lists created by `binary_to_list` should not
  be considered as "strings". This conversion is NOT Unicode-aware, so lists and
  binaries have byte-wise correspondence.
- On the other hand, `unicode:characters_to_list` is Unicode-aware, so converted
  lists are not necessarily `[0..255]`. Each element of a converted list is
  correspond to Unicode codepoint (not so precise but it's almost a character).
- The talk [1] is good reference for Latin-1 and Unicode in Erlang.
  Codepoints, encodings, string literals, binaries and more.

```
> A = <<16#E3, 16#81, 16#82>>. % A binary of Japanese Hiragana "A" in UTF-8
<<227,129,130>>
> binary_to_list(A).
[227,129,130]                  % Just byte-wise
> unicode:characters_to_list(A).
[12354]                        % A list of single element (single character)
```

[1] http://www.erlang-factory.com/conference/ErlangUserConference2013/speakers/PatrikNyblom

### CODE REVIEW

**Comment by shino** [comment]
memo: This PR has conflict with #791. (almost trivial to merge)

**Comment by reiddraper** [comment]
All tests passing. Reviewing the code now.

**Comment by reiddraper** [comment]
Double-checking my Erlang and unicode: is this true? For-all utf-8 encoded binaries, `B`:

``` erlang
B =:= list_to_binary(binary_to_list(B)).
```

File: riak_test/tests/access_stats_test.erl [hunk 1]
```diff
@@ -66,46 +68,69 @@ flush_access_stats() ->
...
-                                      end, ListOfNodeStats),
-    rtcs:json_get(<<"Samples">>, NodeStats).
+    NodeBin = list_to_binary(Node),
+    [NodeStats | _] = lists:dropwhile(
+                        fun(NodeStats) ->
+                                rtcs:json_get(<<"Node">>, NodeStats) =/= NodeBin
+                        end, ListOfNodeStats),
+    rtcs:json_get(<<"Samples">>, NodeStats);
+node_samples_from_content(xml, Node, Content) ->
+    {Usage, _Rest} = xmerl_scan:string(unicode:characters_to_list(Content, utf8)),
```
Comment (reiddraper) [inline riak_test/tests/access_stats_test.erl:108]: In a lot of the other code, we simply `binary_to_list/1` UTF-8 encoded binaries into lists. Does the `xmerl` API demand we give it a list of character points instead?

**Comment by reiddraper** [comment]
I can't think of 'a better way', but one thing that is confusing (and already exists in the codebase) is the mixed use of binaries (presumably UTF-8 encoded), lists (constructed from `binary_to_list/1`) and lists of character points. Is there any way to be more consistent, or are we doing the best we can already?

File: riak_test/tests/user_test.erl [hunk 2]
```diff
@@ -34,38 +34,52 @@ confirm() ->
...
     pass.
 
-user_listing_json_test_case(UserCreds, UserConfig, Node) ->
-    user_listing_test(UserCreds, UserConfig, Node, ?JSON).
+japanese_aiueo() ->
+    %% To avoid dependency on source code encoding, create list from chars.
+    %% These five numbers represents "あいうえお" (A-I-U-E-O in Japanese).
+    %% unicode:characters_to_binary([12354,12356,12358,12360,12362]).
+    Chars = [12354,12356,12358,12360,12362],
+    binary_to_list(unicode:characters_to_binary(Chars)).
```
Comment (reiddraper) [inline riak_test/tests/user_test.erl:61]: Personal note: this is equal to `<<"あいうえお"/utf8>>`.

Comment (reiddraper) [inline riak_test/tests/user_test.erl:61]: Minus the `binary_to_list` part.

File: src/riak_cs_xml.erl [hunk 3]
```diff
@@ -222,7 +245,7 @@ format_value(Val) when is_binary(Val) ->
 format_value(Val) when is_integer(Val) ->
     integer_to_list(Val);
```
Comment (reiddraper) [inline src/riak_cs_xml.erl:246]: nice, this is much cleaner than scattering `integer_to_list` around the codebase.

**Comment by shino** [comment]
> Double-checking my Erlang and unicode: is this true? For-all utf-8 encoded binaries, B:

True (or I believe it's true).

File: riak_test/tests/user_test.erl (hunk 2 above)
Comment (shino) [inline riak_test/tests/user_test.erl:61]: I must use it! Will fix.

**Comment by shino** [comment]
> Does the xmerl API demand we give it a list of character points instead?

xmerl's document says it accepts binary as element of deep list (direct binary is not accepted).
Sample code of list of code points and binary > https://gist.github.com/shino/3c5ab84980dde4e7ab24
It seems xmerl converts a binary to byte-wise list (same as `binary_to_list`).

File: riak_test/tests/user_test.erl (hunk 2 above)
Comment (reiddraper) [inline riak_test/tests/user_test.erl:61]: Well, as your comment mentions, your code doesn't depend on UTF-8 encoding in the actual file, which might be desirable. That being said, I assume most (all?) of us are using text-editors that use UTF-8.

**Comment by shino** [comment]
One trial branch : https://github.com/basho/riak_cs/compare/release%2F1.4...bugfix%2Fnon-ascii-in-xml-2
Difference from `bugfix/non-ascii-in-xml` is only this commit: https://github.com/basho/riak_cs/commit/6a13a83ee74e0d478ed563b93c84860999d1d9e9

No Unicode code points appears in generating XML.
The trick behind is that  `xmerl:export_simple` convert list of binaries to b2l'd lists.
(I wish it accepted direct binary()).

Anyway, XML generation is now encapsulated in `riak_cs_xml`, so switching
these two ways (or another?) is rather easy.

File: riak_test/tests/user_test.erl (hunk 2 above)
Comment (shino) [inline riak_test/tests/user_test.erl:61]: Yeah, editor's default encoding is UTF-8 :)
Another file encoding issue is about compiler. Erlang compiler thinks files are Latin-1 encoded (until R15, also in R16 but can be overwritten magic comments on files).

`"あ"` is three-byte in utf-8 encoded file, compiler sees it as if latin-1 and convert to three latin-1 characters. Then applies `/utf8` conversion so it results a binary of six bytes (oops).

**Comment by shino** [comment]
Because of conflict with #789, rebased and created new PR #796 .

**Comment by shino** [comment]
(reply to myself)

Avoiding `unicode` module at `bugfix/non-ascii-in-xml-2` seems not good for me...
`unicode` can be avoided in generating XML by `xmerl:export_simple` but parsing XML
by `xmerl_scan:string` returns text value of Unicode code points. So they are not symmetric. :see_no_evil:

**Comment by shino** [comment]
Continue in #796 or #807

**Comment by shino** [comment]
memo: #807  is finally merged to release branch `release/1.4`.

**Pull request closed without merging**